File: 2007-find-original-array-from-doubled-array/2007-find-original-array-from-doubled-array.cs
public class Solution {
    public int[] FindOriginalArray(int[] changed) {
            Array.Sort(changed);

            if (changed.Length % 2 != 0) return new int[0];

            int mid = changed.Length / 2;

            int[] res = new int[mid];

            int[] freq = new int[100001];

            foreach (var item in changed)
            {
                freq[item]++;
            }

            int idx = 0;

            foreach (int no in changed)
            {
                if (freq[no] > 0 && no * 2 <= 100000 && freq[no * 2] > 0)
                {
                    freq[no]--;
                    freq[no * 2]--;
                    res[idx++] = no;
                }
            }

            for (int i = 0; i < freq.Length; i++)
            {
                if (freq[i] != 0) return new int[0];
            }

            return res;
    }
}

File: 2007-find-original-array-from-doubled-array/2007-find-original-array-from-doubled-array.cpp
class Solution {
public:
    vector<int> findOriginalArray(vector<int>& changed) {
        map<int,int> m;
        vector<int> res;
        
        if(changed.size()%2) return res;

        sort(changed.begin(),changed.end());
        
        for (auto &&i : changed)
        {
            m[i]++;
        }
        for (auto &&i : changed)
        {
            if(m[i]){
                m[i]--;
                res.push_back(i);
                if(m[2*i]){
                    m[2*i]--;
                }else{
                    return {};
                }
            }
        }
        return res;
    }
};